omnimove_external_control/src/omnimove_external_control.cpp

```diff
@@ -0,0 +1,357 @@
+    command_interface.emplace_back("pillar1", "position", &position_commands_[0]);
+    command_interface.emplace_back("pillar2", "position", &position_commands_[1]);
+    command_interface.emplace_back("pillar3", "position", &position_commands_[2]);
+    command_interface.emplace_back("pillar4", "position", &position_commands_[3]);
+    command_interface.emplace_back("shield", "position", &position_commands_[4]);
+  }
+  else
+  {
+    command_interface.emplace_back("move_x", "velocity", &velocity_commands_[0]);
+    command_interface.emplace_back("move_y", "velocity", &velocity_commands_[1]);
+    command_interface.emplace_back("move_theta", "velocity", &velocity_commands_[2]);
+  }
+
+  return command_interface;
+}
+
+ExternalControlData OmnimoveExternalControl::parseLastMessageFromBuffer()
+{
+  // assume that sequence will start from the beginning
+  const int expected_data_size = ExternalControlData::totalMessageLength();
+
+  size_t last_message_start = (read_buffer_.size() / expected_data_size - 1) * expected_data_size;
+  if (last_message_start > read_buffer_.size())
+  {
+    RCLCPP_ERROR(
+      rclcpp::get_logger("OmnimoveExternalControl"),
+      "Message too small last_message_start %lu, read_buffer_size %lu", last_message_start,
+      read_buffer_.size());
+    return ExternalControlData();
+  }
+  //        RCLCPP_INFO(rclcpp::get_logger("OmnimoveExternalControl"),
+  //                     "last_message_start %lu, read_buffer_size %lu", last_message_start,
+  //                     read_buffer_.size ());
+
+  for (size_t i = 0; i < last_message_start; ++i)
+  {
+    read_buffer_.pop_front();
+  }
+  // now let's check the header.
+  std::unique_ptr<char[]> msg_data = std::make_unique<char[]>(read_buffer_.size());
+  for (size_t i = 0; i < read_buffer_.size(); ++i)
+  {
+    msg_data.get()[i] = read_buffer_[i];
+  }
+  if (!ExternalControlData::isMessageValid(msg_data.get(), read_buffer_.size()))
+  {
+    const char * EXTERNAL_CONTROL_DATA_HEADER = "KMRUTV03";
+
+    for (size_t i = 0; i < strlen(EXTERNAL_CONTROL_DATA_HEADER); ++i)
+    {
+      if (read_buffer_[i] != EXTERNAL_CONTROL_DATA_HEADER[i])
+      {
+        RCLCPP_ERROR(
+          rclcpp::get_logger("OmnimoveExternalControl"), "msg_data is %d, header is %d",
+          read_buffer_[i], EXTERNAL_CONTROL_DATA_HEADER[i]);
+      }
+    }
+
+    return ExternalControlData();
+  }
+  ExternalControlData parsedData(msg_data.get() + 8);
+  read_buffer_.erase(read_buffer_.begin(), read_buffer_.begin() + expected_data_size);
+  return parsedData;
+  // now we will connect the latest message
+
+  // ExternalControlData received_data(read_buffer_.);
+}
+hardware_interface::return_type OmnimoveExternalControl::read(
+  const rclcpp::Time &, const rclcpp::Duration &)
+{
+  std::array<char, 1024> buffer;
+  if (client_socket_.get() == NULL)
+  {
+    RCLCPP_WARN(rclcpp::get_logger("OmnimoveExternalControl"), "socket is null !!!");
```

Comment:
I think this is a bigger issue then a warning message, you should return an error

Reply:
done